casanovo/denovo/dataloaders.py

```diff
@@ -237,46 +235,80 @@ def prepare_batch(
-        (ii) the precursor information [mass, charge, m/z], (iii) the
-        peptide sequence, the precursor m/z, (iv) spectrum identifiers
-        (file and scan).
+        spectrum (i) a tensor with the m/z and intensity peak values, (ii), the
+        precursor m/z, (iii) the precursor charge, (iv) the spectrum identifier.
+    protein_database : db_utils.ProteinDatabase
+        The protein database to use for candidate peptide retrieval.
 
     Returns
     -------
-    spectra : torch.Tensor of shape (batch_size, n_peaks, 2)
+    all_spectra : torch.Tensor of shape (batch_size, n_peaks, 2)
         The padded mass spectra tensor with the m/z and intensity peak values
         for each spectrum.
-    precursors : torch.Tensor of shape (batch_size, 3)
+    all_precursors : torch.Tensor of shape (batch_size, 3)
         A tensor with the precursor neutral mass, precursor charge, and
         precursor m/z.
-    spectrum_peps : np.ndarray
-        Peptide sequences
-    spectrum_ids : Tuple[str, str]
-        Peak file and spectrum identifier
+    all_spectrum_ids : np.ndarray
+        The spectrum identifiers.
+    all_peptides : List[str]
+        The candidate peptides for each spectrum.
+    all_proteins : List[str]
+        The proteins associated with each candidate peptide.
     """
-    (
-        spectra,
-        precursor_mzs,
-        precursor_charges,
-        spectrum_peps,
-        spectrum_ids,
-    ) = list(zip(*batch))
+    spectra, precursor_mzs, precursor_charges, spectrum_ids = list(zip(*batch))
     spectra = torch.nn.utils.rnn.pad_sequence(spectra, batch_first=True)
+
     precursor_mzs = torch.tensor(precursor_mzs)
     precursor_charges = torch.tensor(precursor_charges)
     precursor_masses = (precursor_mzs - 1.007276) * precursor_charges
     precursors = torch.vstack(
         [precursor_masses, precursor_charges, precursor_mzs]
     ).T.float()
-    return spectra, precursors, np.asarray(spectrum_peps), spectrum_ids
+
+    all_spectra = []
+    all_precursors = []
+    all_spectrum_ids = []
+    all_peptides = []
+    all_proteins = []
+    for idx in range(len(batch)):
+        spec_peptides, spec_proteins = protein_database.get_candidates(
+            precursor_mzs[idx].type(torch.float64).item(),
+            precursor_charges[idx].type(torch.int64).item(),
+        )
+        try:
+            all_spectra.append(
+                spectra[idx].unsqueeze(0).repeat(len(spec_peptides), 1, 1)
+            )
+            all_precursors.append(
+                precursors[idx].unsqueeze(0).repeat(len(spec_peptides), 1)
+            )
+            all_spectrum_ids.extend([spectrum_ids[idx]] * len(spec_peptides))
+            all_peptides.extend(spec_peptides)
+            all_proteins.extend(spec_proteins)
+        except ValueError:
+            logger.warning(
+                "No candidates found for spectrum %s", spectrum_ids[idx]
+            )
+            continue
```

Comment:
**nitpick:** `continue` is redundant here.

Reply:
Removed redundant "continue"